packages/core/src/protocol.ts

```diff
@@ -11,22 +11,41 @@ export type LnurlpRequest = {
   /** ReceiverAddress is the address of the user at VASP2 that is receiving the payment. */
   receiverAddress: string;
   /** Nonce is a random string that is used to prevent replay attacks. */
-  nonce: string;
+  nonce?: string | undefined;
   /** Signature is the base64-encoded signature of sha256(ReceiverAddress|Nonce|Timestamp). */
-  signature: string;
+  signature?: string | undefined;
   /** IsSubjectToTravelRule indicates VASP1 is a financial institution that requires travel rule information. */
-  isSubjectToTravelRule: boolean;
+  isSubjectToTravelRule?: boolean | undefined;
   /** VaspDomain is the domain of the VASP that is sending the payment. It will be used by VASP2 to fetch the public keys of VASP1. */
-  vaspDomain: string;
+  vaspDomain?: string | undefined;
   /** Timestamp is the unix timestamp of when the request was sent. Used in the signature. */
-  timestamp: Date;
+  timestamp?: Date | undefined;
   /**
    * The version of the UMA protocol that VASP2 has chosen for this transaction based on its own support and VASP1's specified preference in the LnurlpRequest.
    * For the version negotiation flow, see https://static.swimlanes.io/87f5d188e080cb8e0494e46f80f2ae74.png
    */
-  umaVersion: string;
+  umaVersion?: string | undefined;
 };
 
+export function isLnurlpRequestForUma(
+  request: LnurlpRequest,
+): request is LnurlpRequest & {
```

Comment:
btw, I didn't know about this type guard thing in typescript before this, but it's very nice for tipping off the compiler for what these functions mean. Saved a lot of type assertions in the demo vasp :-D

Reply:
Yeah these are one of the handier TypeScript features to learn! We have a lot of them in our other TS projects. Even dynamic type guards like [these](https://github.com/lightsparkdev/js-sdk/blob/main/packages/ui/src/types/utils.ts#L19), although, I'm not going to lie, I have no idea how this even works. ChatGPT does though!